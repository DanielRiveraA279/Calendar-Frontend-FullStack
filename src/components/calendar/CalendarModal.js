import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import moment from 'moment';
import Modal from 'react-modal';
import DateTimePicker from 'react-datetime-picker';
import Swal from 'sweetalert2';
import { uiCloseModal } from '../../actions/ui';
import { eventStartAddNew, eventClearActiveEvent, eventStartUpdate } from '../../actions/events';

const customStyles = {
    content: {
        top: '50%',
        left: '50%',
        right: 'auto',
        bottom: 'auto',
        marginRight: '-50%',
        transform: 'translate(-50%, -50%)',
    },
};

//si estamos en test esto no se ejecuta
if (process.env.NODE_ENV !== 'test') {
    // Make sure to bind modal to your appElement (https://reactcommunity.org/react-modal/accessibility/)
    Modal.setAppElement('#root');
}


const now = moment().minutes(0).seconds(0).add(1, 'hours'); //si son las 23:26 se va a poner en 00:00
const nowPlusOne = now.clone().add(1, 'hours');

const initEvent = {
    title: '',
    notes: '',
    start: now.toDate(),
    end: nowPlusOne.toDate()
}

export const CalendarModal = () => {

    const { modalOpen } = useSelector(state => state.ui);
    const { activeEvent } = useSelector(state => state.calendar)
    const dispatch = useDispatch();

    const [dateStart, setDateStart] = useState(now.toDate());
    const [dateEnd, setDateEnd] = useState(nowPlusOne.toDate());
    const [titleValid, setTitleValid] = useState(true);

    const [formValues, setFormValues] = useState(initEvent);

    const { notes, title, start, end } = formValues;


    useEffect(() => {
        if (activeEvent) {
            setFormValues(activeEvent);
        } else {
            setFormValues(initEvent); //si activeEvent esta en null, vuevlve todo a 0 por defecto
        }

    }, [activeEvent, setFormValues]);


    const handleInputChange = ({ target }) => {
        setFormValues({
            ...formValues,
            [target.name]: target.value
        })
    }

    const closeModal = () => {
        // TODO: cerrar el modal
        dispatch(uiCloseModal());
        dispatch(eventClearActiveEvent()); //limpiamos los datos seleccionados para el show modal
        setFormValues(initEvent);
    }

    const handleStartDateChange = (e) => {
        setDateStart(e); //recibimos la fecha seleccionada de inicio
        setFormValues({
            ...formValues,
            start: e
        });
    }

    const handleEndDateChange = (e) => {
        setDateEnd(e); //recibimos la fecha seleccionada de fin
        setFormValues({
            ...formValues,
            end: e
        });
    }

    const handleSubmitForm = (e) => {
        e.preventDefault();

        const momentStart = moment(start);
        const momentEnd = moment(end);

        if (momentStart.isSameOrAfter(momentEnd)) {
            Swal.fire('Error', 'La fecha fin debe de ser mayor a la fecha de inicio', 'error');
            return;
        }

        if (title.trim().length < 2) {
            return setTitleValid(false);
        }

        if (activeEvent) {
            dispatch(eventStartUpdate(formValues)) //actualizar un registro
        } else {
            //TODO: agregar un nuevo registro
            dispatch(eventStartAddNew(formValues));
        }



        setTitleValid(true);
        closeModal();
    }

    return (
        <Modal
            isOpen={modalOpen}
            // onAfterOpen={afterOpenModal}
            onRequestClose={closeModal}
            style={customStyles}
            closeTimeoutMS={200} //tiempo en cerrarse el modal
            className='modal'
            overlayClassName='modal-fondo'
            ariaHideApp={!process.env.NODE_ENV === 'test'}
        >
            <h1> {(activeEvent) ? 'Editar Event' : 'Nuevo Evento'} </h1>
            <hr />
            <form
                className="container"
                onSubmit={handleSubmitForm}

            >

                <div className="form-group">
                    <label>Fecha y hora inicio</label>
                    <DateTimePicker
                        onChange={handleStartDateChange}
                        value={dateStart}
                        className='form-control'
                    />
                    {/* <input className="form-control" placeholder="Fecha inicio" /> */}
                </div>

                <div className="form-group">
                    <label>Fecha y hora fin</label>
                    <DateTimePicker
                        onChange={handleEndDateChange}
                        value={dateEnd}
                        minDate={dateStart}
                        className='form-control'
                    />
                    {/* <input className="form-control" placeholder="Fecha inicio" /> */}
                </div>

                <hr />
                <div className="form-group">
                    <label>Titulo y notas</label>
                    <input
                        type="text"
                        className={`form-control ${!titleValid && 'is-invalid'}`}
                        placeholder="Título del evento"
                        name="title"
                        autoComplete="off"
                        value={title}
                        onChange={handleInputChange}
                    />
                    <small id="emailHelp" className="form-text text-muted">Una descripción corta</small>
                </div>

                <div className="form-group">
                    <textarea
                        type="text"
                        className="form-control"
                        placeholder="Notas"
                        rows="5"
                        name="notes"
                        value={notes}
                        onChange={handleInputChange}
                    ></textarea>
                    <small id="emailHelp" className="form-text text-muted">Información adicional</small>
                </div>

                <button
                    type="submit"
                    className="btn btn-outline-primary btn-block"
                >
                    <i className="far fa-save"></i>
                    <span> Guardar</span>
                </button>

            </form>
        </Modal>
    )
};
