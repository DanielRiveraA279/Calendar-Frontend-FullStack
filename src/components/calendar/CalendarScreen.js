import React, { useEffect, useState } from 'react';
import { Calendar, momentLocalizer } from 'react-big-calendar';
import moment from 'moment';

import { NavBar } from '../ui/NavBar';
import { messages } from '../../helpers/calendar.messages.es';
import { CalendarEvent } from './CalendarEvent';
import { CalendarModal } from './CalendarModal';

import 'moment/locale/es'; //configuracion para las semanas en espanol
import 'react-big-calendar/lib/css/react-big-calendar.css';
import { useDispatch, useSelector } from 'react-redux';
import { uiOpenModal } from '../../actions/ui';
import { eventClearActiveEvent, eventSetActive, eventStartLoading } from '../../actions/events';
import { AddNewFab } from '../ui/AddNewFab';
import { DeleteEventFab } from '../ui/DeleteEventFab';

//config para espanol
moment.locale('es');

//configurar momentLocalizer para que muestre el calendar
const localizer = momentLocalizer(moment);

// const events = [
//     {
//         title: 'Cumple del jefe',
//         start: moment().toDate(),
//         end: moment().add(2, 'hours').toDate(),
//         bgcolor: '#fafafa',
//         notes: 'Comprar el pastel',
//         user: {
//             _id: '1234',
//             name: 'Daniel Antonio Rivera'
//         }
//     }
// ]

export const CalendarScreen = () => {

    const dispatch = useDispatch();
    const { events, activeEvent } = useSelector(state => state.calendar);
    const { uid } = useSelector(state => state.auth);

    //TODO: leer 

    const [lastView, setLastView] = useState(localStorage.getItem('lastView') || 'month');

    useEffect(() => {
        dispatch(eventStartLoading());
    }, [dispatch]);

    //evento para doble click
    const onDoubleClick = (e) => {
        dispatch(uiOpenModal());
    }

    //evento para un solo click
    const onSelectEvent = (e) => {
        // console.log(e);
        dispatch(eventSetActive(e));
    }

    //evento para saber en que vista estamos ejemp. mes, semana, dia, etc
    const onViewChange = (e) => {
        setLastView(e); //actualizamos con el ultimo cambio de vista que tiene
        localStorage.setItem('lastView', e); //grabamos la vista actual
    }

    //consigue la info o data que seleccione
    const eventStyleGetter = (event, start, end, isSelected) => {
        const style = {
            backgroundColor: (uid === event.user._id) ? '#367CF7' : '#465660',
            borderRadius: '0px',
            opacity: 0.8,
            display: 'block',
            color: 'white'
        }

        return {
            style
        }
    }

    const onSelectSlot = (e) => {
        // console.log(e)
        dispatch(eventClearActiveEvent()); //para que se limpie el state, cada vez que clickeamos a fuera de una nota
    }

    return (
        <div className='calendar-screen'>
            <NavBar />

            <Calendar
                localizer={localizer}
                events={events}
                startAccessor="start"
                endAccessor="end"
                messages={messages}
                eventPropGetter={eventStyleGetter}
                onSelectEvent={onSelectEvent}
                onDoubleClickEvent={onDoubleClick}
                onView={onViewChange}
                onSelectSlot={onSelectSlot}
                selectable={true}
                view={lastView}
                components={{
                    event: CalendarEvent
                }}
            />

            <AddNewFab />

            {
                (activeEvent) && <DeleteEventFab />
            }

            <CalendarModal />
        </div>
    );
};
