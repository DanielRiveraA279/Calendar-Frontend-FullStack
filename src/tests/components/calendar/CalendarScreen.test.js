import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';

import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import '@testing-library/jest-dom';
import { act } from '@testing-library/react';

import { CalendarScreen } from '../../../components/calendar/CalendarScreen';
import { messages } from '../../../helpers/calendar.messages.es'
import { types } from '../../../types/types';
import { eventSetActive, eventStartLoading } from '../../../actions/events';

jest.mock('../../../actions/events', () => ({
    eventSetActive: jest.fn(),
    eventStartLoading: jest.fn(),
}));

Storage.prototype.setItem = jest.fn(); //mock de localStorage

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {
    calendar: {
        events: []
    },
    auth: {
        uid: '123',
        name: 'Daniel'
    },
    ui: {
        openModal: false
    }
};

const store = mockStore(initState);
//manejamos moock
store.dispatch = jest.fn();

//Armar la estructura de redireccion para el componente <DeleteEventFab />
const wrapper = mount(
    <Provider store={store}>
        <CalendarScreen />
    </Provider>
)

describe('Pruebas en <CalendarScreen>', () => {

    test('debe de mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot();
    })

    test('pruebas con las interacciones del calendario', () => {
        const calendar = wrapper.find('Calendar'); //buscamos el componente <Calendar />
        const calendarMessages = calendar.prop('messages') //nos posicionamos en la prop messages del componente <Calendar />

        expect(calendarMessages).toEqual(messages) //comparamos los valores de messages

        calendar.prop('onDoubleClickEvent')(); //posicionamos en la prop

        //y esperamos que se ejecute un dispatch y que llame a este type
        expect(store.dispatch).toHaveBeenCalledWith({ type: types.uiOpenModal });

        calendar.prop('onSelectEvent')({ start: 'Hola' }) //se le pasa ese valor por defecto a la prop
        expect(eventSetActive).toHaveBeenCalledWith({ start: 'Hola' }) //se espera que este evento tenga el mismo objeto asignado


        //act: action
        act(() => {
            calendar.prop('onView')('week');
            expect(localStorage.setItem).toHaveBeenCalledWith('lastView', 'week');
        })


    })
})