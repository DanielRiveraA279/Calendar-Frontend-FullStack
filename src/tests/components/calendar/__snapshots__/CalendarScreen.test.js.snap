// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <CalendarScreen> debe de mostrarse correctamente 1`] = `ReactWrapper {}`;
